sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/RewriteWithExpression.scala

```diff
@@ -35,59 +35,50 @@ object RewriteWithExpression extends Rule[LogicalPlan] {
   override def apply(plan: LogicalPlan): LogicalPlan = {
     plan.transformWithPruning(_.containsPattern(WITH_EXPRESSION)) {
       case p if p.expressions.exists(_.containsPattern(WITH_EXPRESSION)) =>
-        val commonExprs = mutable.ArrayBuffer.empty[Alias]
-        // `With` can be nested, we should only rewrite the leaf `With` expression, as the outer
-        // `With` needs to add its own Project, in the next iteration when it becomes leaf.
-        // This is done via "transform down" and check if the common expression definitions does not
-        // contain nested `With`.
-        var newPlan: LogicalPlan = p.transformExpressionsDown {
-          case With(child, defs) if defs.forall(!_.containsPattern(WITH_EXPRESSION)) =>
-            val idToCheapExpr = mutable.HashMap.empty[Long, Expression]
-            val idToNonCheapExpr = mutable.HashMap.empty[Long, Alias]
-            defs.zipWithIndex.foreach { case (commonExprDef, index) =>
+        var newChildren = p.children
+        var newPlan: LogicalPlan = p.transformExpressionsUp {
+          case With(child, defs) =>
+            val refToExpr = mutable.HashMap.empty[Long, Expression]
+            val childProjections = Array.fill(newChildren.size)(mutable.ArrayBuffer.empty[Alias])
+
+            defs.foreach(commonExprDef =>
               if (CollapseProject.isCheap(commonExprDef.child)) {
-                idToCheapExpr(commonExprDef.id) = commonExprDef.child
+                refToExpr(commonExprDef.id) = commonExprDef.child
+              } else {
+                val index = newChildren.indexWhere(
+                  c => commonExprDef.child.references.subsetOf(c.outputSet)
+                )
+                if (index == -1) {
+                  // When we cannot rewrite the common expressions, force to inline them so that the
+                  // query can still run. This can happen if the join condition contains `With` and
+                  // the common expression references columns from both join sides.
+                  // TODO: things can go wrong if the common expression is nondeterministic. We
+                  //       don't fix it for now to match the old buggy behavior when certain
+                  //       `RuntimeReplaceable` did not use the `With` expression.
+                  // TODO: we should calculate the ref count and also inline the common expression
+                  //       if it's ref count is 1.
+                  refToExpr(commonExprDef.id) = commonExprDef.child
+                } else {
+                  val alias = Alias(commonExprDef.child, s"_common_expr_${commonExprDef.id}")()
```

Comment:
The reason I chose the expr definition index is to make the logical plan stable. Spark Connect has gold files test that record the logical plan. We can normalize the alias name there, but I think it's simpler to just don't use the def id here.

Reply:
I see and makes sense. Fixed in https://github.com/cloud-fan/spark/pull/19/commits/50ecc13ee49f6d1c73958dd1d40dd166b713235a.